src/renderer/components/shared/CanvasDataBuilder.js

```diff
@@ -528,12 +571,17 @@ const buildRPInstances = async (answers, currentData, shouldLimit, graknTx) => {
               if (shouldLimit) rpEntries = rpEntries.slice(0, QuerySettings.getNeighboursLimit());
               rpEntries.forEach(([role, players], i) => {
                 role.label().then((edgeLabel) => {
-                  players.forEach((player) => {
+                  players.forEach((player, j) => {
                     player.type().then(type => type.label().then((playerLabel) => {
                       player.label = playerLabel;
-                      edges.push(getEdge(relation, player, edgeTypes.instance.RELATES, edgeLabel));
-                      nodes.push(getInstanceNode(player, graqlVar, answer.explanation, answer.queryPattern));
-                      if (i === rpEntries.length - 1) resolve({ edges, nodes });
+                      const edge = getEdge(relation, player, edgeTypes.instance.RELATES, edgeLabel);
+                      getInstanceNode(player, graqlVar, answer.explanation, answer.queryPattern).then((node) => {
+                        edges.push(edge);
+                        nodes.push(node);
+                        const isLastRole = i === rpEntries.length - 1;
+                        const isLastPlayer = j === players.length - 1;
+                        if (isLastRole && isLastPlayer) resolve({ edges, nodes });
```

Comment:
this is awks oof but i dont think there's an alternative is there huh

Reply:
I'm sure this code can be made more readable, I just can't see how at this moment. perhaps, my next week's studying will help me rewrite this in a cleaner way :)

I've created this issue (https://github.com/graknlabs/workbase/issues/285) for it to not be forgotten

Reply:
yeah i saw a stackoverflow that basically said if you're using async stuff in a foreach its all weird and not guaranteed to work if you put the call at the outside of the forEach or something weird, idk :P